src/ui/map.js

```diff
@@ -324,6 +325,22 @@ class Map extends Camera {
         this._hash = options.hash && (new Hash()).addTo(this);
         // don't set position from options if set through hash
         if (!this._hash || !this._hash._onHashChange()) {
+            if (options.bounds) {
+                const bounds = LngLatBounds.convert(options.bounds),
+                    tr = this.transform,
+                    nw = tr.project(bounds.getNorthWest()),
+                    se = tr.project(bounds.getSouthEast()),
+                    size = se.sub(nw),
+                    dimensions = this._containerDimensions(),
+                    width = dimensions[0],
+                    height = dimensions[1],
+                    scaleX = width / size.x,
+                    scaleY = height / size.y;
+
+                options.center = tr.unproject(nw.add(se).div(2));
+                options.zoom = Math.min(tr.scaleZoom(tr.scale * Math.min(scaleX, scaleY)), options.maxZoom);
+            }
+
```

Comment:
Can we reuse the code from `fitBounds` somehow and avoid copying it?

Reply:
Actually, we can try to use `fitBounds(bounds, { duration: 0 })` instead of manually reproducing `fitBounds` logic, but we have to resize the map before.

```js
this.resize();
this.fitBounds(options.bounds, { duration: 0 });
```

Unfortunately, the problem with the test is still there.

Reply:
I have updated the PR using `fitBounds`